tests/react/utils/types.test.tsx

```diff
@@ -0,0 +1,16 @@
+import { it } from '@jest/globals'
+import { useHydrateAtoms } from 'jotai/react/utils'
+import { atom } from 'jotai/vanilla'
+
+it('useHydrateAtoms should not allow invalid atom types when array is passed', () => {
+  function Component() {
+    const countAtom = atom(0)
+    const activeAtom = atom(true)
+    // @ts-expect-error TS2769
```

Comment:
Actually this is the wrong `ts-expect-error` 😞, ideally the test would be:
```tsx
it('useHydrateAtoms should not allow invalid atom types when array is passed', () => {
  function Component() {
    const countAtom = atom(0)
    const activeAtom = atom(true)
    useHydrateAtoms([
      // @ts-expect-error TS2322
      [countAtom, 'foo'],
      // @ts-expect-error TS2322
      [activeAtom, 0],
    ])
    useHydrateAtoms([
      [countAtom, 1],
      // @ts-expect-error TS2322
      [activeAtom, 0],
    ])
    useHydrateAtoms([
      // @ts-expect-error TS2322
      [countAtom, true],
      [activeAtom, false],
    ])
  }
  Component
})
```
but the `Map` overloading is preventing type script from correctly picking `Array`/`Iterable` over `T extends Map`:
https://github.com/pmndrs/jotai/blob/6b03f946edb41d4dcb70bc9686886820c4390e6c/src/react/utils/useHydrateAtoms.ts#L19-L22

I'm at a loss because the above test would be the most preferred and correct.